src/url-loader/indirect-url-resolver.ts

```diff
@@ -0,0 +1,122 @@
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {FileRelativeUrlBrand, PackageRelativeUrl} from '../index';
+import {FileRelativeUrl, ResolvedUrl} from '../model/url';
+
+import {FsUrlResolver} from './fs-url-resolver';
+
+type RuntimeUrl = string&RuntimeUrlBrand;
+
+
+// Declare these as classes rather than interfaces so that the properties
+// can be private.
+declare class RuntimeUrlBrand extends FileRelativeUrlBrand {
+  private RuntimeUrlBrand: never;
+}
+
+/**
+ * A URL resolver for very large codebases where source files map in an
+ * arbitrary but predetermined fashion onto URL space.
+ *
+ * It also separates the root directory – the root of all source code that's
+ * legal to load – from the package directory, which is how the user refers to
+ * files on the CLI or the IDE.
+ */
+export class IndirectUrlResolver extends FsUrlResolver {
+  private readonly runtimeUrlToResolvedUrl:
+      ReadonlyMap<RuntimeUrl, ResolvedUrl>;
+  private readonly resolvedUrlToRuntimeUrl:
+      ReadonlyMap<ResolvedUrl, RuntimeUrl>;
+
+  /**
+   * @param rootPath All loadable source code must be a descendent of this
+   *     directory. Should be the same as FsUrlLoader.
+   * @param packagePath The base directory for paths from the user. Usually the
+   *     current working directory.
+   * @param urlToFilesystemMap Maps the runtime URL space to the paths for those
+   *     files on the filesystem. The runtime URLs should all be relative paths
+   *     from the same base url. The filesystem paths should all be relative
+   *     paths from `rootPath`.
+   */
+  constructor(
+      rootPath: string, packagePath: string,
+      indirectionMap: Map<string, string>) {
+    super(packagePath);
+
+    const rootResolver = new FsUrlResolver(rootPath);
+    const urlspaceToFilesystem = new Map<RuntimeUrl, ResolvedUrl>();
+    const filesystemToUrlspace = new Map<ResolvedUrl, RuntimeUrl>();
+    for (const [u, fsPath] of indirectionMap) {
+      const url = u as RuntimeUrl;
+      const fsUrl = rootResolver.resolve(this.brandAsRelative(fsPath));
+      if (fsUrl === undefined) {
+        throw new Error(`Invalid fs path in indirection map: ${fsPath}`);
+      }
+      urlspaceToFilesystem.set(url, fsUrl);
+      filesystemToUrlspace.set(fsUrl, url);
+    }
+    this.runtimeUrlToResolvedUrl = urlspaceToFilesystem;
+    this.resolvedUrlToRuntimeUrl = filesystemToUrlspace;
+  }
+
+  resolve(url: FileRelativeUrl|PackageRelativeUrl, baseUrl?: ResolvedUrl):
+      ResolvedUrl|undefined {
+    // If we're just resolving user input, do the basic thing.
+    if (baseUrl === undefined) {
+      return super.resolve(url);
+    }
+    const runtimeBaseUrl = this.resolvedUrlToRuntimeUrl.get(baseUrl);
+    if (runtimeBaseUrl === undefined) {
+      return super.resolve(url, baseUrl);
```

Comment:
why is it ok to delegate to super here? If we can't handle the base, we don't we return `undefined`?